### PR TITLE
Feature request/delete how tos and research articles

### LINKED ISSUE #2391: [feature request] allow user to delete their posts
**Is your feature request related to a problem? Please describe.**
User adds post (how-to / research) in error or realises it's a duplicate and would like to delete post and is unable to.

**Describe the solution you'd like**
Addition of a delete button that is visible to the user that created the post on how to and research articles.

**Describe alternatives you've considered**
Manually finding the item in the database and deleting it.

**Additional context**
[Discord request](https://discord.com/channels/586676777334865928/1108116011191902218/1111559755169800292)

### PR DESCRIPTION
**Pull request by KungRaseri**
PR Checklist

- [x] - Commit [messages are descriptive](https://github.com/ONEARMY/community-platform/blob/master/CONTRIBUTING.md#--commit-style-guide), it will be used in our [Release Notes](https://github.com/ONEARMY/community-platform/releases/)

PR Type

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Developer experience (improves developer workflows for contributing to the project)

## Description

- [x] Add a Delete button to the How-To description section, next to the Edit button.
  - [x] Soft-delete with `_deleted` property set to true
  - [x] Filter deleted items out of how-tos list
  - [x] When viewing soft-deleted how-to, the delete button is disabled 
  - [x] When viewing soft-deleted how-to, "* Marked for deletion" is clearly visible at the top of the section
- [x] Add a Delete button to the Research description section, next to the Edit button.
  - [x] Soft-delete with `_deleted` property set to true
  - [x] Filter deleted items out of research list
  - [x] When viewing soft-deleted research, the delete button is disabled 
  - [x] When viewing soft-deleted research, "* Marked for deletion" is clearly visible at the top of the section
- [x] Add test coverage
  - [x] Unit testing (partial)
  - [x] E2E testing (partial)

## Git Issues

Closes #2391 

### How-Tos:
The Button:
image
The Modal:
image
The Aftermath!
image

### Research
The Button!
image
The Modal!
image
The Aftermath!
image

### CODE REVIEW

File: src/pages/Howto/Content/Howto/HowtoDescription/HowtoDescription.tsx [hunk 1]
```diff
@@ -88,6 +98,23 @@ const HowtoDescription = ({ howto, loggedInUser, ...props }: IProps) => {
     }
   }
 
+  const handleDelete = async (_id: string) => {
+    await stores.howtoStore.deleteHowTo(_id)
```
Comment (thisislawatts) [inline src/pages/Howto/Content/Howto/HowtoDescription/HowtoDescription.tsx:102]: **issue:** The `deleteHowTo` method you've added may throw an error so you may want to catch the error here.

**Comment by KungRaseri** [comment]
Added a discord discussion thread [here](https://discord.com/channels/586676777334865928/1120511589741776916) with a few questions.

**Comment by KungRaseri** [comment]
Questions: 
1. Should we redirect the user to `/how-to` and `/research` respectively after the delete action is completed?
2. Should we "cascade delete" all comments associated with a how-to/research article when it is deleted?
3. Should we limit the ability to delete to `author` and `super-admin`, or also include `admin`?

**Comment by thisislawatts** [comment]
@KungRaseri thanks for raising those questions 💜 

1. Should we redirect the user to /how-to and /research respectively after the delete action is completed?
That seems like a reasonable approach to me :) 

2. Should we "cascade delete" all comments associated with a how-to/research article when it is deleted?
I think we should be soft deleting how-to/research articles by adding a `_deleted:true` property to items which a User has opted to remove this will mean that:
* A platform admin or moderater can undelete in case of user error.
* Hard deleting of articles,how-tos and comments can be managed as an occasional background task. For example after an item has been marked as deleted for more than 30 days.

3. Should we limit the ability to delete to author and super-admin, or also include admin?
The delete capability should be available for the original author, then admin and super-admin roles.

**Comment by KungRaseri** [comment]
> @KungRaseri thanks for raising those questions 💜
> 
>     1. Should we redirect the user to /how-to and /research respectively after the delete action is completed?
>        That seems like a reasonable approach to me :)
> 
>     2. Should we "cascade delete" all comments associated with a how-to/research article when it is deleted?
>        I think we should be soft deleting how-to/research articles by adding a `_deleted:true` property to items which a User has opted to remove this will mean that:
> 
> 
>     * A platform admin or moderater can undelete in case of user error.
> 
>     * Hard deleting of articles,how-tos and comments can be managed as an occasional background task. For example after an item has been marked as deleted for more than 30 days.
> 
> 
>     3. Should we limit the ability to delete to author and super-admin, or also include admin?
>        The delete capability should be available for the original author, then admin and super-admin roles.

Great! Thanks for the reply!  Just a few follow ups for now:
1. With the `_deleted` property, would it be beneficial or necessary to add a `_deletedBy` property as well?
2. Should we setup another issue for the background task? or should we try to get that setup with this feature request?

**Comment by thisislawatts** [comment]
> 1. With the _deleted property, would it be beneficial or necessary to add a _deletedBy property as well?

I think that we should emit this data on the tracking event and log it, persisting it to the document seems unnecessary. Is there a specific scenario you had in mind that wouldnt be solved by event tracking/logs? 

> 2. Should we setup another issue for the background task? or should we try to get that setup with this feature request?

No need for this at the moment, this seems like a future work item but I can't see when it would be done. So I think we trust that when/if it becomes important again someone will flag it.

**Comment by AlfonsoGhislieri** [comment]
Just jumping in here @KungRaseri to say make sure you are using prettier as this is required to pass linting tests! 


https://prettier.io/

**Comment by KungRaseri** [comment]
> Just jumping in here @KungRaseri to say make sure you are using prettier as this is required to pass linting tests!
> 
> https://prettier.io/

Thanks for the reminder! I had it all setup, but didn't get to actually running it :)

I did have quite a few updates from it for files I haven't touched, which I'm sure is fine.  Just wanted to give a heads up.

There also seemed to be an issue in the `pk-template.handlebars` file that I resolved.

**Comment by KungRaseri** [comment]
This is ready for final review 👍 Thanks for all the feedback!

File: .github/pr-labeler.config.yml
Comment (thisislawatts) [inline .github/pr-labeler.config.yml:1]: **question:**  @KungRaseri what command did you use to format the project with prettier? Using `yarn format` will apply both prettier and eslint to files it looks like you have updated some files here which should not be updated. It may be useful to tidy up for project wide consistency at a later date but it introduces unnecessarily noise to the this PR. Would you be able to drop this commit from the branch?

File: functions/src/emailNotifications/pk-template.handlebars [hunk 2]
```diff
@@ -52,7 +52,8 @@
                   {{this.triggeredBy.displayName}}
```
Comment (thisislawatts) [inline functions/src/emailNotifications/pk-template.handlebars:52]: **issue (blocking):** Please move this commit to a separate branch and raise a separate PR. This looks like a useful fix but it needs to be reviewed in isolation as it is unrelated to deleting how-tos and research articles.

File: .circleci/config.yml
Comment (thisislawatts) [inline .circleci/config.yml:1]: **Praise:** Thanks for introducing this tooling @KungRaseri, really interested to see what we can learn from these reports!

I have split this out to [a standalone PR](https://github.com/ONEARMY/community-platform/pull/2460) so it can be reviewed in isolation. Please drop [this commit](799e76776a4fe70dccb05949b5d3f2422a400a2d) from the branch.

File: src/stores/Howto/howto.store.tsx [hunk 3]
```diff
@@ -468,6 +470,41 @@ export class HowtoStore extends ModuleStore {
...
+      const howToData = await dbRef.get('server')
+      const user = this.activeUser
+
+      if (id && howToData && user) {
+        const updatedHowTo: IHowtoDB = {
+          ...howToData,
+          _deleted: true,
+        }
+
+        await dbRef.set(
```
Comment (thisislawatts) [inline src/stores/Howto/howto.store.tsx:487]: **Issue (blocking):** Please use the `updateHowtoItem`, passing through the `{...howToData, _deleted: true}` argument.

File: src/stores/Research/research.store.test.ts [hunk 4]
```diff
@@ -586,6 +586,21 @@ describe('research.store', () => {
...
+      // Act
+      await store.deleteResearch(researchItem._id)
+
+      // Assert
+      const [deletedResearchItem] = setFn.mock.calls[0]
+      expect(setFn).toHaveBeenCalledTimes(1)
+      expect(getFn).toHaveBeenCalledTimes(1)
+      expect(deletedResearchItem._deleted).toBeTruthy()
+    })
+  })
```
Comment (thisislawatts) [inline src/stores/Research/research.store.test.ts:602]: **suggestion:** Could we introduce a test that validate the behaviour for one of the failure scenarios. For example if the `set` fails?

File: src/stores/Research/research.store.tsx [hunk 5]
```diff
@@ -252,8 +258,41 @@ export class ResearchStore extends ModuleStore {
...
+
+      const user = this.activeUser
+
+      if (id && researchData && user) {
+        const updatedResearch: IResearchDB = {
+          ...researchData,
+          _deleted: true,
+        }
+
+        await dbRef.set(
```
Comment (thisislawatts) [inline src/stores/Research/research.store.tsx:275]: **issue:** Similar to the howto store can you use the `_updateResearchItem` method to update the underlying document?  We want to ensure the fewest possible pathways in updating documents to make it easier to introduce hooks around saving/retrieving documents. Apologies this could be better documented within the code itself.

**Comment by AlfonsoGhislieri** [comment]
Also @KungRaseri it would be preferable if instead of merging you could use `rebase` instead. Keeps the commit history a lot cleaner. https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase

Also could use squash to group up commits and avoid a large number of commits. 
https://www.freecodecamp.org/news/git-squash-commits/

**Comment by KungRaseri** [comment]
> Also @KungRaseri it would be preferable if instead of merging you could use `rebase` instead. Keeps the commit history a lot cleaner. https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase
> 
> Also could `squash` to group up commits and avoid a large number of commits. https://www.freecodecamp.org/news/git-squash-commits/

Very new to this part of the process, but I rebased interactively to drop those mentioned commits.  I'll look into it more as we go!

Thank you for the resources!

File: src/stores/Research/research.store.test.ts (hunk 4 above)
Comment (KungRaseri) [reply src/stores/Research/research.store.test.ts:600]: Yeah this is a good idea, I was spacing what to add for that scenario.  I'll think on it a bit more and see what I can come up with.

File: .github/pr-labeler.config.yml
Comment (KungRaseri) [reply .github/pr-labeler.config.yml:1]: I was using `yarn prettier write .`, but I have since started using yarn format instead!

**Comment by thisislawatts** [comment]
@KungRaseri this is looking great, do you have some time to address the merge conflicts? I can take a look at this next week.

**Comment by KungRaseri** [comment]
> @KungRaseri this is looking great, do you have some time to address the merge conflicts? I can take a look at this next week.

Yeah, let's sync up after the 4th!  Gonna be a busy weekend and beginning of the week :D

**Comment by thisislawatts** [comment]
Thanks for all the work on this @KungRaseri, I am going to adopt the changes here, resolve the merge conflicts and push them through to production. I will be sure to leave you as a co-author in the commit history. Thanks again ✨

**Comment by sigolenej** [comment]
Hey @thisislawatts I see this has been closed but has the issue been solved? Are humans now able to delete their HTs? :)
